examples/bond/assets/bond-dapp-stateful.py

```diff
@@ -0,0 +1,206 @@
+        ),
+        Return(Int(1))
+    ])
+
+    # fetch asset_holding.balance from Txn.accounts[0]
+    asset_balance = AssetHolding.balance(Int(1), Gtxn[1].xfer_asset())
+
+    on_create_dex = Seq([
+        asset_balance, # load asset_balance from store
+        Assert(
+            And(
+                Txn.sender() == App.globalGet(store_manager),
+                basic_checks,
+                Gtxn[0].type_enum() == TxnType.AssetTransfer,
+                # transfer `balanceOf(issuer, B_i)`  of `B_{i+1}` from the creator to the `issuer`.
+                # index 1 of Txn.accounts().
+                asset_balance.value() == Gtxn[0].asset_amount(),
+                # burn `B_i` issuer bonds. send to null_address
+                Gtxn[1].type_enum() == TxnType.AssetTransfer,
+                asset_balance.value() == Gtxn[1].asset_amount()
+            ),
+        ),
+        # Increment `BondApp.epoch`
+        App.globalPut(epoch, Add(App.globalGet(epoch), Int(1))),
+        # set `BondApp.current_bond = B_{i+1}`.
+        App.globalPut(current_bond, Gtxn[0].xfer_asset()),
+        Return(Int(1))
+    ])
+
+    program = Cond(
+        # Verfies that the application_id is 0, jumps to on_initialize.
+        [Txn.application_id() == Int(0), on_initialize],
+        # Verifies Update transaction, rejects it.
+        [Txn.on_completion() == OnComplete.UpdateApplication, Return(Int(0))],
+        # Verifies delete transaction, rejects it.
+        [Txn.on_completion() == OnComplete.DeleteApplication, Return(Int(0))],
+        # Verifies closeout transaction, approves it.
+        [Txn.on_completion() == OnComplete.CloseOut, Return(Int(1))],
+        # Verifies opt-in transaction, approves it.
+        [Txn.on_completion() == OnComplete.OptIn, Return(Int(1))],
+        # Verifies update issuer address transaction, jumps to update_issuer branch.
+        [Txn.application_args[0] == Bytes("update_issuer_address"), update_issuer],
+        # Verifies update issue, jumps to on_issue branch.
+        [Txn.application_args[0] == Bytes("issue"), on_issue],
+        # Verifies buy transaction, jumps to on_buy branch.
+        [Txn.application_args[0] == Bytes("buy"), on_buy],
+        # Verifies create buyback transaction, jumps to create_buyback branch.
+        [Txn.application_args[0] == Bytes("create_buyback"), create_buyback],
+        # Verifies exit transaction, jumps to on_exit branch.
+        [Txn.application_args[0] == Bytes("exit"), on_exit],
+        # Verifies redeem coupon transaction, jumps to on_redeem_coupon branch.
+        [Txn.application_args[0] == Bytes("redeem_coupon"), on_redeem_coupon],
+        # Verifies update issue price transaction, jumps to on_update_issue_price branch.
+        [Txn.application_args[0] == Bytes("update_issue_price"), on_update_issue_price],
+        # Verifies create dex transaction, jumps to on_create_dex branch.
+        [Txn.application_args[0] == Bytes("create_dex"), on_create_dex],
+    )
```

Comment:
should we have a default condition - if it doesn't match anything above then reject?

Reply:
we have one assert for store manager check. If store manager doesn't call to create Dex, it will be rejected.
Am I understanding your question correctly ?

Reply:
What happens if someone calls app (noop) with arg: `some-other-argument:str`?

Reply:
it fails with error `logic eval error: TEAL runtime encountered err opcode`

Reply:
ah, so PyTEAL is adding that. Good.